internal/coredns/coredns.go

```diff
@@ -0,0 +1,247 @@
+				logger.Info("CoreDNS not found, skipping CoreDNS config patch")
+				return nil
+			}
+			return err
+		}
+
+		// patches contain each key and value that needs to be patched in the coredns configmap data field.
+		patches, err := generatePatches(kubeClient, overrides, isK3d)
+		if err != nil {
+			return err
+		}
+		if len(patches) != 0 {
+			cm, err = doPatch(logger, kubeClient, patches)
+			return err
+		}
+		return nil
+	}, retry.Delay(2*time.Second), retry.Attempts(3), retry.DelayType(retry.FixedDelay))
+
+	return
+}
+
+func doPatch(logger *zap.Logger, kubeClient kubernetes.Interface, patches map[string]string) (cm *v1.ConfigMap, err error) {
+	configMaps := kubeClient.CoreV1().ConfigMaps("kube-system")
+	coreDNSConfigMap, err := configMaps.Get(context.TODO(), "coredns", metav1.GetOptions{})
+	exists := true
+	if err != nil {
+		if apierr.IsNotFound(err) {
+			exists = false
+		} else {
+			return cm, err
+		}
+	}
+
+	if exists {
+		logger.Info("Patching CoreDNS config")
+		return patchCoreDNSConfigMap(logger, configMaps, coreDNSConfigMap, patches)
+	}
+
+	logger.Info("Corefile not found, creating new CoreDNS config")
+	return createCoreDNSConfigMap(logger, configMaps, patches)
+}
+
+func patchCoreDNSConfigMap(logger *zap.Logger, configMaps corev1.ConfigMapInterface, coreDNSConfigMap *v1.ConfigMap, patch map[string]string) (cm *v1.ConfigMap, err error) {
+	for k, v := range patch {
+		coreDNSConfigMap.Data[k] = v
+	}
+	jsontext, err := json.Marshal(coreDNSConfigMap)
+	if err != nil {
+		return
+	}
+
+	cm, err = configMaps.Patch(context.TODO(), "coredns", types.StrategicMergePatchType, jsontext, metav1.PatchOptions{})
+	if err != nil {
+		logger.Error("Could not patch CoreDNS config")
+	}
+	return
+}
+
+func createCoreDNSConfigMap(logger *zap.Logger, configMaps corev1.ConfigMapInterface, patch map[string]string) (cm *v1.ConfigMap, err error) {
+	cm, err = configMaps.Create(context.TODO(), newCoreDNSConfigMap(patch), metav1.CreateOptions{})
+	if err != nil {
+		logger.Error("Could not create new CoreDNS config")
+	}
+	return
+}
+
+func newCoreDNSConfigMap(data map[string]string) *v1.ConfigMap {
+	return &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{Name: "coredns"},
+		Data:       data,
+	}
+}
+
+func generatePatches(kubeClient kubernetes.Interface, overrides overrides.Overrides, isK3s bool) (map[string]string, error) {
+	patches := make(map[string]string)
+	// patch the CoreFile only if not on gardener and no custom domain is provided
+	gardenerDomain, err := gardener.Domain(kubeClient)
+	if err != nil {
+		return nil, err
+	}
+	if err != nil {
+		return nil, err
+	}
+	_, hasCustomDomain := overrides.Find("global.domainName")
+	if gardenerDomain == "" && !hasCustomDomain {
+		var domainName string
+		if isK3s {
+			domainName = coreDNSLocalDomainName
+		} else {
+			domainName = coreDNSRemoteDomainName
+		}
+		patches["Corefile"], err = generateCorefile(domainName)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
I think it would be better to first check if its a K3d cluster, before trying to fetch the Gardener domain, to save up some calls.
Furthermore, we changed the naming from `K3s` to `K3d` in most part of Kyma (also in CLI command), thus we should also change it in this new command to `K3d`.

Reply:
Agree, but since it's untested functionality (there are literally 0 tests) we just moved from the parallel-install pkg, I'd not refactor the logic. I'll apply the renaming.